test/grouping.jl

```diff
@@ -1218,4 +1238,126 @@ end
     end
 end
 
+@testset "Dictionary interface" begin
+    df = DataFrame(a = repeat([:A, :B, missing], outer=4), b = repeat([:X, :Y], inner=6), c = 1:12)
+    gd = groupby_checked(df, [:a, :b])
+
+    @test map(NamedTuple, keys(gd)) ≅
+        [(a=:A, b=:X), (a=:B, b=:X), (a=missing, b=:X), (a=:A, b=:Y), (a=:B, b=:Y), (a=missing, b=:Y)]
+
+    @test collect(pairs(gd)) ≅ map(Pair, keys(gd), gd)
+
+    for (i, key) in enumerate(keys(gd))
+        # Plain key
+        @test gd[key] ≅ gd[i]
+        # Named tuple
+        @test gd[NamedTuple(key)] ≅ gd[i]
+        # Plain tuple
+        @test gd[Tuple(key)] ≅ gd[i]
+    end
+
+    @test get(gd, (a=:A, b=:X), nothing) ≅ gd[1]
+    @test get(gd, (a=:A, b=:Z), nothing) == nothing
+
+    @test_throws KeyError gd[(a=:A, b=:Z)]
+    @test_throws KeyError gd[(a=:A, b="Z")]
+    @test_throws KeyError gd[(a=:A,)]
+    @test_throws KeyError gd[(a=:A, b=:X, c=1)]
+    @test_throws KeyError gd[(:A, :Z)]
+    @test_throws KeyError gd[(:A,)]
+    @test_throws KeyError gd[(:A, :X, 1)]
+    @test_throws KeyError gd[(b=:X, a=:D)]
+end
+
+@testset "GroupKey" begin
+    df = DataFrame(a = repeat([:A, :B, missing], outer=4), b = repeat([:X, :Y], inner=6), c = 1:12)
+    cols = [:a, :b]
+    gd = groupby_checked(df, cols)
+    gdkeys = keys(gd)
+
+    # Check type
+    N = Tuple(cols)
+    V = Tuple{(eltype(df[!, c]) for c in cols)...}
+    K = DataFrames.GroupKey{N, V}
+    @test eltype(gdkeys) === K
+
+    expected =
+        [(a=:A, b=:X), (a=:B, b=:X), (a=missing, b=:X), (a=:A, b=:Y), (a=:B, b=:Y), (a=missing, b=:Y)]
+
+    # Test each key
+    for (i, key) in enumerate(gdkeys)
+        nt = expected[i]
+
+        # Basic methods
+        @test parent(key) === gd
+        @test length(key) == length(cols)
+        @test names(key) == N
+        @test keys(key) == N
+        @test values(key) ≅ values(nt)
+
+        # (Named)Tuple conversion
+        @test Tuple(key) ≅ values(nt)
+        @test NamedTuple(key) ≅ nt
+
+        # Iteration
+        @test collect(key) ≅ collect(nt)
+
+        # Integer/symbol indexing, getproperty of key
+        for (j, n) in enumerate(cols)
+            @test key[j] ≅ nt[j]
+            @test key[n] ≅ nt[j]
+            @test getproperty(key, n) ≅ nt[j]
+        end
+
+        # Invalid key/property of key
+        @test_throws KeyError key[:foo]
+        @test_throws ErrorException key.foo
+
+        # Using key to index GroupedDataFrame
+        @test gd[key] ≅ gd[i]
+    end
+
+    # Indexing using another GroupedDataFrame instance should fail
+    gd2 = groupby(df, cols)
+    @test_throws ErrorException gd2[gdkeys[1]]
+
+    # Key equality
+    @test keys(gd) == gdkeys  # These are new instances
+    @test all(Ref(gdkeys[1]) .!= gdkeys[2:end])
+    @test !any(keys(gd2) .== gdkeys)  # Same values but different parent
+
+    # Type stability for value access methods
+    @test Base.return_types(values, (K,)) == [V]
+    # _get1(x) = x[1]
```

Comment:
?

Reply:
This is testing that the compiler can correctly infer that `values(::GroupKey{N,V})` returns the type V. Typically you would use `Test.@inferred` but it doesn't work here because V is `Tuple{Union{Missing, Symbol}, Symbol}` but the actual return type will be one of the two concrete subtypes of this.

Reply:
I guess the question is why do you need the comment in line 1331.

Reply:
Ah, these were similar tests for accessing elements of the key by index or property name. These would also be inferred correctly for a NamedTuple, but I left them commented out because I haven't implemented that yet. I'm not sure if its worth it.